service/storage_proxy.cc

```diff
@@ -210,6 +210,36 @@ class storage_proxy::remote {
 
     bool _stopped{false};
 
+    // Transform a read command by reversing the schema version and the bounds of the clustering key ranges. The transformation
+    // is done only for reverse read queries in a mixed node cluster when the native_reverse_queries feature flag is turned off.
+    query::read_command transform_to_from_legacy_format_for_native_reversed_queries(query::read_command cmd, bool reversed)
+    {
+        if (!reversed || _sp.features().native_reverse_queries)
+            return cmd;
+
+        auto schema = local_schema_registry().get(cmd.schema_version)->get_reversed();
+        cmd.schema_version = schema->version();
+        cmd.slice = query::legacy_reverse_slice_to_native_reverse_slice(*schema, cmd.slice);
+
+        return cmd;
+    }
+
+    // Transform a reconcilable_result by reversing stored mutations. The transformation is done only for reverse read queries
+    // in a mixed node cluster when the native_reverse_queries feature flag is turned off.
+    reconcilable_result transform_to_from_legacy_format_for_native_reversed_queries(reconcilable_result result, bool reversed)
+    {
+        if (!reversed || _sp.features().native_reverse_queries)
+            return result;
```

Comment:
_sp.feature checks should only be done on the client side. On the server side, only check the command.

Otherwise you are subject to the following race:

client: advertise feature
server: advertise feature
client: recognize cluster has feature
client: send request that relies on the feature
server: mishandle request
server: recognize cluster has feature